server/plugin/google/google.go

```diff
@@ -0,0 +1,348 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &DocsService{
+		service: srv,
+		googleServiceBase: googleServiceBase{
+			serviceType: docsServiceType,
+			papi:        g.papi,
+			limiter:     nil,
+			userID:      userID,
+			kvstore:     g.kvstore,
+		},
+	}, nil
+}
+
+func (g *Client) NewSlidesService(ctx context.Context, userID string) (*SlidesService, error) {
+	authToken, err := g.getGoogleUserToken(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = checkKVStoreLimitExceeded(g.kvstore, slidesServiceType, userID)
+	if err != nil {
+		return nil, err
+	}
+
+	srv, err := slides.NewService(ctx, option.WithTokenSource(g.oauthConfig.TokenSource(ctx, authToken)))
+	if err != nil {
+		return nil, err
+	}
+
+	return &SlidesService{
+		service: srv,
+		googleServiceBase: googleServiceBase{
+			serviceType: slidesServiceType,
+			papi:        g.papi,
+			limiter:     nil,
+			userID:      userID,
+			kvstore:     g.kvstore,
+		},
+	}, nil
+}
+
+func (g *Client) NewSheetsService(ctx context.Context, userID string) (*SheetsService, error) {
+	authToken, err := g.getGoogleUserToken(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = checkKVStoreLimitExceeded(g.kvstore, sheetsServiceType, userID)
+	if err != nil {
+		return nil, err
+	}
+
+	srv, err := sheets.NewService(ctx, option.WithTokenSource(g.oauthConfig.TokenSource(ctx, authToken)))
+	if err != nil {
+		return nil, err
+	}
+
+	return &SheetsService{
+		service: srv,
+		googleServiceBase: googleServiceBase{
+			serviceType: sheetsServiceType,
+			papi:        g.papi,
+			limiter:     nil,
+			userID:      userID,
+			kvstore:     g.kvstore,
+		},
+	}, nil
+}
+
+func (g *Client) NewDriveActivityService(ctx context.Context, userID string) (*DriveActivityService, error) {
+	authToken, err := g.getGoogleUserToken(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = checkKVStoreLimitExceeded(g.kvstore, driveActivityServiceType, userID)
+	if err != nil {
+		return nil, err
+	}
+
+	srv, err := driveactivity.NewService(ctx, option.WithTokenSource(g.oauthConfig.TokenSource(ctx, authToken)))
+	if err != nil {
+		return nil, err
+	}
+
+	return &DriveActivityService{
+		service: srv,
+		googleServiceBase: googleServiceBase{
+			serviceType: driveActivityServiceType,
+			papi:        g.papi,
+			limiter:     nil,
+			userID:      userID,
+			kvstore:     g.kvstore,
+		},
+	}, nil
+}
+
+func (g *Client) getGoogleUserToken(userID string) (*oauth2.Token, error) {
+	encryptedToken, err := g.kvstore.GetGoogleUserToken(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(encryptedToken) == 0 {
+		return nil, nil
+	}
+
+	decryptedToken, err := utils.Decrypt([]byte(g.config.EncryptionKey), string(encryptedToken))
+	if err != nil {
+		return nil, err
+	}
+
+	var oauthToken oauth2.Token
+	err = json.Unmarshal([]byte(decryptedToken), &oauthToken)
+
+	return &oauthToken, err
+}
+
+func (ds googleServiceBase) parseGoogleErrors(ctx context.Context, apiErr error) error {
```

Comment:
Is `ctx` used here?

Reply:
I think I could name this method better. The method is used to check for rate limit errors and if rate limit errors are found it will take the appropriate steps. If no rate limit errors are found it will return the original error. 

There is nowhere in the method where it will return nil, even if it finds a rate limit error it still returns the original error. If there is an error saving to the kvstore it will also wrap that error with it.

I think I'll rename this to something along the lines of `checkForRateLimitErrors`

Reply:
https://github.com/mattermost-community/mattermost-plugin-google-drive/pull/25/commits/7ed12e1d806c23d47914a8b50c7e7fb11ca2e425 renamed the method and removed context